web/src/stream_data.ts

```diff
@@ -267,20 +278,33 @@ export function slug_to_name(slug: string): string {
     */
     const m = /^(\d+)(-.*)?$/.exec(slug);
     if (m) {
-        const stream_id = Number.parseInt(m[1]!, 10);
-        const sub = sub_store.get(stream_id);
-        if (sub) {
-            return sub.name;
+        const stream_id_string = m[1]!;
+        const stream_id = Number.parseInt(stream_id_string, 10);
+        let stream = stream_info.get(stream_id);
+        if (stream) {
+            return stream_id_string;
         }
         // if nothing was found above, we try to match on the stream
         // name in the somewhat unlikely event they had a historical
         // link to a stream like 4-horsemen
+        const stream_name = m[2]!.slice(1); // Remove the "-" at the start
+        stream = get_sub_by_name(stream_name);
+        if (stream) {
+            return stream.stream_id.toString();
+        }
+        // TODO(evy): what should we do here?
+        return slug;
```

Comment:
I don't think you need any new code here. The case discussed in this comment is that where the stream name is "4-horsemen", which might happen to match the pattern for modern URLs. The intended behavior is to fall through to the logic below for handling the pre-2018 URLs, which do not start with the stream ID.

(This fallthrough was about pre-2018 URLs that appear to be a modern URL because the stream name starts with the pattern `4-`)

Reply:
So you're saying remove this whole block?